client/block.c

```diff
@@ -921,74 +923,97 @@ int xdag_create_block(struct xdag_field *fields, int inputsCount, int outputsCou
 	}
 
 	if (mining) {
-		uint64_t taskIndex = g_xdag_pool_task_index + 1;
-		struct xdag_pool_task *task = &g_xdag_pool_task[taskIndex & 1];
-
-		xdag_generate_random_array(block[0].field[XDAG_BLOCK_FIELDS - 1].data, sizeof(xdag_hash_t));
+		if(!do_mining(block, pretop, send_time)) {
+			goto begin;
+		}
+	}
 
-		task->task_time = MAIN_TIME(send_time);
+	xdag_hash(block, sizeof(struct xdag_block), newBlockHash);
 
-		xdag_hash_init(task->ctx0);
-		xdag_hash_update(task->ctx0, block, sizeof(struct xdag_block) - 2 * sizeof(struct xdag_field));
-		xdag_hash_get_state(task->ctx0, task->task[0].data);
-		xdag_hash_update(task->ctx0, block[0].field[XDAG_BLOCK_FIELDS - 2].data, sizeof(struct xdag_field));
-		memcpy(task->ctx, task->ctx0, xdag_hash_ctx_size());
+	if(mining) {
+		memcpy(g_xdag_mined_hashes[MAIN_TIME(send_time) & (CONFIRMATIONS_COUNT - 1)],
+			newBlockHash, sizeof(xdag_hash_t));
+		memcpy(g_xdag_mined_nonce[MAIN_TIME(send_time) & (CONFIRMATIONS_COUNT - 1)],
+			block[0].field[XDAG_BLOCK_FIELDS - 1].data, sizeof(xdag_hash_t));
+	}
 
-		xdag_hash_update(task->ctx, block[0].field[XDAG_BLOCK_FIELDS - 1].data, sizeof(struct xdag_field) - sizeof(uint64_t));
-		memcpy(task->task[1].data, block[0].field[XDAG_BLOCK_FIELDS - 2].data, sizeof(struct xdag_field));
-		memcpy(task->nonce.data, block[0].field[XDAG_BLOCK_FIELDS - 1].data, sizeof(struct xdag_field));
-		memcpy(task->lastfield.data, block[0].field[XDAG_BLOCK_FIELDS - 1].data, sizeof(struct xdag_field));
+	log_block("Create", newBlockHash, block[0].field[0].time, 1);
+	
+	if(block_hash_result != NULL) {
+		memcpy(block_hash_result, newBlockHash, sizeof(xdag_hash_t));
+	}
 
-		xdag_hash_final(task->ctx, &task->nonce.amount, sizeof(uint64_t), task->minhash.data);
-		g_xdag_pool_task_index = taskIndex;
+	struct xdag_block *new_block = (struct xdag_block *)malloc(sizeof(struct xdag_block));
```

Comment:
Could you do an error check? :)

Reply:
Ok, but in my opinion we have to close the pool in case of insufficient memory.

Reply:
Done